### DISCUSSION TITLE
Dashboard with oled 1306 display issues

### DISCUSSION THREAD

**Discussion by Stephano2000** [davetcc/tcMenu#305] [Q&A]
Short of having a TFT color display handy atm, I converted the Adafruit dashboard example to use the oled display using the U8G2 display plugin,  MonoOled theme with TcUnicode checked and fonts set to Adafruit RobotoMedium18 (setting them to TcUnicode RobotoMedium18 yields duplication errors). As soon as i enter the settings submenu, the display just freezes until the OK button is pressed and we're back to the menu. I even trimmed the dashboard, removed the delegate altogether, and kept a single drawing item with the same results.
When the delegate is setup, its functions are indeed being called (WillDraw and DidDraw repetitively).

Any pointers about what is wrong?
Is there a debug setting/define to trace what's going on?

TcMenu ver 3.1.0 / ESP8266 / Ubuntu 22.04

**Comment by Stephano2000** [comment]
I even tried a bare bones example from scratch:

DrawableDashboard* mainDashboard;
...
  SetupDashboard();
...
void CALLBACK_FUNCTION OnTest1Click(int id) {
  renderer.takeOverDisplay();
}

void SetupDashboard() {
  mainDashboard = new DrawableDashboard(renderer.getDeviceDrawable(), &renderer, nullptr,
                                        DrawableDashboard::DASH_ON_RESET_CLICK_EXIT);
  mainDashboard->addDrawingItem(&menuTest2, Coord(30, 0), nullptr, 10, nullptr, 10);
  renderer.setCustomDrawingHandler(mainDashboard);
}

The display doesn't freeze, but nothing happens in this case

**Reply by davetcc** [reply]
The dashboard has not been tried on anything except for TFTs, it probably is not ending the drawing transaction, which pushes the pixels.

Further, at the moment, I've not really tried tcUnicode on smaller displays, I found the fonts looked slightly out. I have a plan to fix that in an upcoming release. It may be easier to allow dashboards that don't need tcUnicode.

The dashboard and tcUnicode support is very new, and will take a release or two to settle in (thats why tcUnicode is off by default).

**Reply by davetcc** [reply]
>  It may be easier to allow dashboards that don't need tcUnicode.

I may do a library patch that allows for dashboards with native fonts.

**Reply by Stephano2000** [reply]
> > It may be easier to allow dashboards that don't need tcUnicode.
> 
> I may do a library patch that allows for dashboards with native fonts.

The barebones example above doesn't use TcUnicode. It uses the defaut font option (U8G2).

**Reply by davetcc** [reply]
Unfortunately, though when you use a dash it turns it on forcibly, that's why the dashboard failed. That's what the patch would be for

**Reply by Stephano2000** [reply]
> Unfortunately, though when you use a dash it turns it on forcibly, that's why the dashboard failed. That's what the patch would be for

Please do the library patch the soonest as the dashboard seems to be the main and easy way to provide a custom display.

Thank you

**Reply by davetcc** [reply]
See https://github.com/davetcc/tcMenuLib/issues/186

**Comment by Stephano2000** [comment]
Using the tcMenuLib and tcUnicodeHelper master branch, I converted the Adafruit dashboard example to use the oled display using the U8G2 display plugin, MonoOled theme with TcUnicode checked and fonts set to Adafruit RobotoMedium24. There were some compilation errors, maybe due to using the 3.1.0 menu editor:

/home/Arduino-ESP-Projects/picoAdafruitDashboard/dashboardConfig.cpp: In member function 'virtual bool MyDrawableDashboardDelegate::dashboardWillOpen(BaseMenuRenderer*)':
/home/Arduino-ESP-Projects/picoAdafruitDashboard/dashboardConfig.cpp:127:9: error: 'gfxDrawable' was not declared in this scope; did you mean 'U8g2Drawable'?
  127 |         gfxDrawable.setDrawColor(LCD_LIGHT_BLUE);
      |         ^~~~~~~~~~~
      |         U8g2Drawable
/home/Arduino-ESP-Projects/picoAdafruitDashboard/picoAdafruitDashboard_menu.cpp:19:41: error: invalid conversion from 'int' to 'const u8g2_cb_t*' {aka 'const u8g2_cb_struct*'} [-fpermissive]
   19 | U8G2_SSD1306_128X64_NONAME_F_HW_I2C gfx(1, 19, -1, -1);
      |                                         ^
      |                                         |
      |                                         int
In file included from /home/Arduino/libraries/tcUnicodeHelper/src/tcUnicodeHelper.h:91,
                 from /home/Arduino-ESP-Projects/picoAdafruitDashboard/picoAdafruitDashboard_menu.h:16,
                 from /home/Arduino-ESP-Projects/picoAdafruitDashboard/picoAdafruitDashboard_menu.cpp:12:
/home/Arduino/libraries/U8g2/src/U8g2lib.h:1687:64: note:   initializing argument 1 of 'U8G2_SSD1306_128X64_NONAME_F_HW_I2C::U8G2_SSD1306_128X64_NONAME_F_HW_I2C(const u8g2_cb_t*, uint8_t, uint8_t, uint8_t)'
 1687 |   public: U8G2_SSD1306_128X64_NONAME_F_HW_I2C(const u8g2_cb_t *rotation, uint8_t reset = U8X8_PIN_NONE, uint8_t clock = U8X8_PIN_NONE, uint8_t data = U8X8_PIN_NONE) : U8G2() {
      |                                               ~~~~~~~~~~~~~~~~~^~~~~~~~

exit status 1

Compilation error: 'gfxDrawable' was not declared in this scope; did you mean 'U8g2Drawable'?

**Reply by davetcc** [reply]
Ah, that's probably because you're u8g2 variable is probably not called "gfxDrawable", the drawable will be the variable of type U8g2Drawable and that will be exported in the proejctName_menu.h file.

To be honest, I would copy the bits of code out of the example that you need, changing from Adafruit to U8G2 is quite a big change.

**Reply by davetcc** [reply]
Although I haven't got a dashboard on an OLED via U8G2, there should be nothing that particularly stops it from working, and I know the core is working on U8G2 as I'm testing card layout with it now.

**Reply by Stephano2000** [reply]
In the U8G2 case, gfxDrawable is defined in xxx_menu.cpp:
U8G2_SSD1306_128X64_NONAME_F_HW_I2C gfx(1, 19, -1, -1);
U8g2Drawable gfxDrawable(&gfx);
GraphicsDeviceRenderer renderer(30, applicationInfo.name, &gfxDrawable);

But not in the xxx_menu.h:
// variables we declare that you may need to access
extern const PROGMEM ConnectorLocalInfo applicationInfo;
extern U8G2_SSD1306_128X64_NONAME_F_HW_I2C gfx;
extern GraphicsDeviceRenderer renderer;
extern const GFXfont RobotoMedium24;

Whereas in the Adafruit case, it is mentioned in xxx_menu.h:
// variables we declare that you may need to access
extern const PROGMEM ConnectorLocalInfo applicationInfo;
extern Adafruit_ILI9341 gfx;
extern AdafruitDrawable gfxDrawable;
extern GraphicsDeviceRenderer renderer;
extern const GFXfont RobotoMedium24;

This is why it might be a Menu Editor issue.

**Reply by davetcc** [reply]
For now just export it yourself using the same method.

**Comment by davetcc** [comment]
the other option is just do "renderer.getDeviceDrawable()" but that returns a pointer. See https://www.thecoderscorner.com/products/arduino-libraries/tc-menu/rendering-with-tcmenu-lcd-tft-oled/#drawing-direct-to-the-display-with-devicedrawable

**Comment by davetcc** [comment]
But the next UI release will export for u8g2 automatically.

**Comment by davetcc** [comment]
See https://github.com/davetcc/tcMenuLib/blob/master/examples/esp/esp32s2Saola/u8g2DashConfig.cpp
